man/cbcscript.8
.TH cbcscript 8 "Version 0.3: 06 May 2020" "CMDB suite manuals" "cmdb, cbc and dnsa collection"
.SH NAME
cbcscript \- Create Build Config Scripts
.SH SYNOPSIS
.B cbcscript
[
.B action
]
[
.B type
]
[
.B specifier
]

.SH DESCRIPTION
\fBcbcscript\fP add scripts to the final stage of a build. These can be used
to configure the completed build, such as creating configuration files for
software installed, run commands on the new build and in other ways configure
the new build. You can add scripts, and then add arguments for these scripts
,specified by a positional number, so multiple arguments can be provided. These
can be static, or they can be dynamic and based on information from the build.
.PP
You will have to group these scripts and options by \fIbuild domain\fP and
\fIbuild type\fP. Build domains are defined in the \fBcbcdomain\fP command,
whereas the build types are one of the base operating systems you can build
for.
.PP
Action and type switches do not take arguments.

.SH OPTIONS
.B actions
.IP "-a,  --add"
add
.IP "-l,  --list"
list
.IP "-d,  --display"
display
.IP "-r,  --remove, --delete"
remove
.PP
.B type
.IP "-f,  --script-arg"
arg
.IP "-s,  --script"
script
.PP
These actions specify what you want to do, and what objects you wish to act
on. How you act on them is specified via the options below, which all take
arguments.
.PP
.B specifiers
.IP "-b,  --domain, --build-domain \fBbuild domain\fP"
specify the build domain
.IP "-o,  --number \fBnumber\fP"
specify the argument number
.IP "-g,  --argument \fBargument\fP"
script argument
.IP "-n,  --name \fBname\fP"
the name of the script
.IP "-t,  --type \fBbuild type\fP"
build type operating system
.PP
.SH EXAMPLES
.B To add a script
.IP
\fBcbcscript -a -s -n \fIscript-name\fR
.PP
.B To add an argument to a script
.IP
\fBcbcscript -a -f -b \fIdomain\fP -n \fIscript-name\fP -t \fIbuild-type-os\fP -g \fIargument\fP -o \fInumber\fR
.PP
.B To list scripts
.IP
\fBcbcscript -l\fP
.PP
.B To display script configs for a domain
.IP
\fBcbcscript -d -b \fIdomain\fR.
.PP
.B To display script configs for a script
.IP
\fBcbcscript -d -n \fIscript-name\fR.
.PP
.B To remove a script argument
.IP
\fBcbcscript -r -f -b \fIdomain\fP -n \fIscript-name\fP -t \fIbuild-type-os\fP -g \fIargument\fR -o \fInumber\fR
.PP
.SH FILES
.I /etc/cmdb/cmdb.conf
.RS
The system wide configuration file for the cmdb / dnsa / cbc suite of
programs. See
.BR cmdb.conf (5)
for further details.
.RE
.I ~/.cmdb.conf
.RS
User configuration for the cmdb / dnsa / cbc suite of programs. See
.BR cmdb.conf (5)
for further details.
.RE
.SH ENVIRONMENT
This suite of programs do not make use of environment variables at present
although this may change in the future. Watch this space!
.SH AUTHOR
Iain M Conochie <iain-at-thargoid-dot-co-dot-uk>
.SH "SEE ALSO"
.BR cbcdomain(8) cbcsysp(8)



NEWS
---- Thursday 04th December 2014 --

Added the --enable-cbc check to configure

---- Wedensday 03rd December 2014 -

Updated dnsa so we can remove glue zones. Also added ability to add a .
at the end of the NS name if it contains the FQDN - otherwise this will
append the domain name to the end.

---- Sunday 30th November 2014 ----

Updated the autoconf configure.ac file - using the m4 directory to store
macros to check for libraries installed. This will be of use when we
start to add more libraries to be linked.

---- Monday 11th February 2012 ----

Release version 0.2.1 contains the new look cmdb. The database interface
has been modularised so that either sqlite or mysql can be used, and
also starting to use prepared statements.
There have been some other code clean ups too.
The 0.2.x branch will concentrate on the modularised structure for 
cbc and dnsa.

---- Friday 25th January 2013 -----

To celebrate Burns night this is the first release of the cmdb programs
using autoconf


src/cbcscript.c
/*
 *
 *  cbcscript: Create Build Configuration Scripts
 *  Copyright (C) 2014 - 2020  Iain M Conochie <iain-AT-thargoid.co.uk>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 *  cbcscript.c
 *
 *  Functions to get configuration values and also parse command line arguments
 *
 *  Part of the cbcscript program
 *
 */
#define _GNU_SOURCE
#include <config.h>
#include <configmake.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <time.h>
#include <unistd.h>
#ifdef HAVE_GETOPT_H
# include <getopt.h>
#endif // HAVE_GETOPT_H
#include <ailsacmdb.h>
#include <ailsasql.h>
#include "cmdb.h"

static int
cbc_script_add_script(ailsa_cmdb_s *cbc, cbc_syss_s *scr);

static int
cbc_script_add_script_arg(ailsa_cmdb_s *cbc, cbc_syss_s *scr);

static int
cbc_script_rm_script(ailsa_cmdb_s *cbc, cbc_syss_s *scr);

static int
cbc_script_rm_arg(ailsa_cmdb_s *cbc, cbc_syss_s *scr);

static int
cbc_script_display_script(ailsa_cmdb_s *cbc, cbc_syss_s *scr);

static int
cbc_script_list_script(ailsa_cmdb_s *cbc);

static int
cbc_script_args_display_all_domain(ailsa_cmdb_s *cbc, cbc_syss_s *scr);

static int
cbc_script_args_display_one_domain(ailsa_cmdb_s *cbc, cbc_syss_s *scr);

static int
cbc_script_args_display_one_script(ailsa_cmdb_s *cbc, cbc_syss_s *scr);

static int
parse_cbc_script_comm_line(int argc, char *argv[], cbc_syss_s *cbcs);

static int
check_cbc_script_comm_line(cbc_syss_s *cbcs);

int
main(int argc, char *argv[])
{
	int retval = 0;
	cbc_syss_s *scr = ailsa_calloc(sizeof(cbc_syss_s), "scr in main");
	ailsa_cmdb_s *cbc = ailsa_calloc(sizeof(ailsa_cmdb_s), "cbc in main");

	if ((retval = parse_cbc_script_comm_line(argc, argv, scr)) != 0) {
		clean_cbc_syss_s(scr);
		ailsa_clean_cmdb(cbc);
		display_command_line_error(retval, argv[0]);
	}
	parse_cmdb_config(cbc);
	if (scr->action == CMDB_ADD) {
		if (scr->what == CBCSCRIPT)
			retval = cbc_script_add_script(cbc, scr);
		else if (scr->what == CBCSCRARG)
			retval = cbc_script_add_script_arg(cbc, scr);
		else
			retval = WRONG_TYPE;
	} else if (scr->action == CMDB_RM) {
		if (scr->what == CBCSCRIPT)
			retval = cbc_script_rm_script(cbc, scr);
		else if (scr->what == CBCSCRARG)
			retval = cbc_script_rm_arg(cbc, scr);
		else
			retval = WRONG_TYPE;
	} else if (scr->action == CMDB_LIST) {
		retval = cbc_script_list_script(cbc);
	} else if (scr->action == CMDB_DISPLAY) {
		retval = cbc_script_display_script(cbc, scr);
	} else {
		retval = WRONG_ACTION;
	}
	ailsa_clean_cmdb(cbc);
	clean_cbc_syss_s(scr);
	if ((retval != 0) && (retval != NO_RECORDS))
		report_error(retval, "");
	return retval;
}

// Helper functions

static int
parse_cbc_script_comm_line(int argc, char *argv[], cbc_syss_s *cbcs)
{
	const char *optstr = "ab:dfg:h:ln:o:rst:v";
	int retval, opt;
	retval = 0;
#ifdef HAVE_GETOPT_H
	int index;
	struct option lopts[] = {
		{"add",			no_argument,		NULL,	'a'},
		{"domain",		required_argument,	NULL,	'b'},
		{"build-domain",	required_argument,	NULL,	'b'},
		{"display",		no_argument,		NULL,	'd'},
		{"script-arg",		no_argument,		NULL,	'f'},
		{"argument",		required_argument,	NULL,	'g'},
		{"help",		no_argument,		NULL,	'h'},
		{"list",		no_argument,		NULL,	'l'},
		{"name",		required_argument,	NULL,	'n'},
		{"number",		required_argument,	NULL,	'o'},
		{"remove",		no_argument,		NULL,	'r'},
		{"delete",		no_argument,		NULL,	'r'},
		{"script",		no_argument,		NULL,	's'},
		{"type",		required_argument,	NULL,	't'},
		{"version",		no_argument,		NULL,	'v'},
		{NULL,			0,			NULL,	0}
	};

	while ((opt = getopt_long(argc, argv, optstr, lopts, &index)) != -1)
#else
	while ((opt = getopt(argc, argv, optstr)) != -1)
#endif // HAVE_GETOPT_H
	{
		if (opt == 'a')
			cbcs->action = CMDB_ADD;
		else if (opt == 'd')
			cbcs->action = CMDB_DISPLAY;
		else if (opt == 'l')
			cbcs->action = CMDB_LIST;
		else if (opt == 'r')
			cbcs->action = CMDB_RM;
		else if (opt == 'h')
			return DISPLAY_USAGE;
		else if (opt == 'v')
			return AILSA_VERSION;
		else if (opt == 's')
			cbcs->what = CBCSCRIPT;
		else if (opt == 'f')
			cbcs->what = CBCSCRARG;
		else if (opt == 'o')
			cbcs->no = strtoul(optarg, NULL, 10);
		else if (opt == 'b')
			cbcs->domain = strndup(optarg, (CONFIG_LEN));
		else if (opt == 'g')
			cbcs->arg = strndup(optarg, (CONFIG_LEN - 1));
		else if (opt == 'n')
			cbcs->name = strndup(optarg, (CONFIG_LEN - 1));
		else if (opt == 't')
			cbcs->type = strndup(optarg, (MAC_LEN - 1));
		else
			retval = DISPLAY_USAGE;
	}
	if (argc == 1)
		retval = DISPLAY_USAGE;
	else
		retval = check_cbc_script_comm_line(cbcs);
	return retval;
}

static int
check_cbc_script_comm_line(cbc_syss_s *cbcs)
{
	int retval = 0;

	if (cbcs->action == 0) {
		retval = NO_ACTION;
	} else if (cbcs->action == CMDB_DISPLAY) {
		if (!(cbcs->domain) && !(cbcs->name))
			retval = NO_DOMAIN_OR_NAME;
	} else if (cbcs->action == CMDB_LIST) {
		cbcs->what = CBCSCRIPT;	// Listing just args makes no sense
	} else if (cbcs->action != CMDB_LIST) {
		if (!(cbcs->name))
			retval = NO_NAME;
		else if (cbcs->what == 0)
			retval = NO_TYPE;
		else if (cbcs->what == CBCSCRARG) {
			if (cbcs->no == 0)
				retval = NO_NUMBER;
			else if (!(cbcs->type))
				retval = NO_ALIAS;
			else if (!(cbcs->domain))
				retval = NO_BUILD_DOMAIN;
			else if (!(cbcs->arg))
				retval = NO_ARG;
		}
	}
	return retval;
}

static int
cbc_script_add_script(ailsa_cmdb_s *cbc, cbc_syss_s *scr)
{
	if (!(cbc) || !(scr))
		return AILSA_NO_DATA;
	AILLIST *l = ailsa_db_data_list_init();
	AILLIST *res = ailsa_db_data_list_init();
	int retval;

	if ((retval = cmdb_add_string_to_list(scr->name, l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add script name to list");
		goto cleanup;
	}
	if ((retval = ailsa_argument_query(cbc, SYSTEM_SCRIPT_ID_ON_NAME, l, res)) != 0) {
		ailsa_syslog(LOG_ERR, "SYSTEM_PACKAGE_ID query failed");
		goto cleanup;
	}
	if (res->total > 0) {
		ailsa_syslog(LOG_INFO, "Script %s already in the database");
		goto cleanup;
	}
	if ((retval = cmdb_populate_cuser_muser(l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add cuser and muser to list");
		goto cleanup;
	}
	if ((retval = ailsa_insert_query(cbc, INSERT_SYSTEM_SCRIPT, l)) != 0)
		ailsa_syslog(LOG_ERR, "INSERT_SYSTEM_SCRIPT query failed");
	cleanup:
		ailsa_list_full_clean(l);
		ailsa_list_full_clean(res);
		return retval;
}

static int
cbc_script_add_script_arg(ailsa_cmdb_s *cbc, cbc_syss_s *scr)
{
	if (!(cbc) || !(scr))
		return AILSA_NO_DATA;
	int retval;
	AILLIST *l = ailsa_db_data_list_init();
	AILLIST *res = ailsa_db_data_list_init();

	if ((retval = cmdb_add_string_to_list(scr->arg, l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add system script arg to list");
		goto cleanup;
	}
	if ((retval = cmdb_add_number_to_list(scr->no, l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add number to list");
		goto cleanup;
	}
	if ((retval = cmdb_add_system_script_id_to_list(scr->name, cbc, l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add system script id to list");
		goto cleanup;
	}
	if ((retval = cmdb_add_build_domain_id_to_list(scr->domain, cbc, l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add build domain id to list");
		goto cleanup;
	}
	if ((retval = cmdb_add_build_type_id_to_list(scr->type, cbc, l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add build type id to list");
		goto cleanup;
	}
	if ((retval = ailsa_argument_query(cbc, SYSTEM_SCRIPT_ARG_ID, l, res)) != 0) {
		ailsa_syslog(LOG_ERR, "SYSTEM_SCRIPT_ARG_ID query failed");
		goto cleanup;
	}
	if (res->total > 0) {
		ailsa_syslog(LOG_INFO, "System script argument is already in the database");
		goto cleanup;
	}
	if ((retval = cmdb_populate_cuser_muser(l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add cuser and muser to list");
		goto cleanup;
	}
	if ((retval = ailsa_insert_query(cbc, INSERT_SYSTEM_SCRIPT_ARGS, l)) != 0)
		ailsa_syslog(LOG_ERR, "INSERT_SYSTEM_SCRIPT_ARGS query failed");
	cleanup:
		ailsa_list_full_clean(l);
		ailsa_list_full_clean(res);
		return retval;
}

// Remove functions

static int
cbc_script_rm_script(ailsa_cmdb_s *cbc, cbc_syss_s *scr)
{
	if (!(cbc) || !(scr))
		return AILSA_NO_DATA;
	int retval;
	AILLIST *l = ailsa_db_data_list_init();
	AILLIST *res = ailsa_db_data_list_init();

	if ((retval = cmdb_add_string_to_list(scr->name, l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add script name to database");
		goto cleanup;
	}
	if ((retval = ailsa_argument_query(cbc, SYSTEM_SCRIPT_ID_ON_NAME, l, res)) != 0) {
		ailsa_syslog(LOG_ERR, "SYSTEM_SCRIPT_ID_ON_NAME query failed");
		goto cleanup;
	}
	if (res->total == 0) {
		ailsa_syslog(LOG_INFO, "System script %s not found in the database");
		goto cleanup;
	}
	if ((retval = ailsa_delete_query(cbc, delete_queries[DELETE_SYSTEM_SCRIPT], res)) != 0)
		ailsa_syslog(LOG_ERR, "DELETE_SYSTEM_SCRIPT query failed");

	cleanup:
		ailsa_list_full_clean(l);
		ailsa_list_full_clean(res);
		return retval;
}

static int
cbc_script_rm_arg(ailsa_cmdb_s *cbc, cbc_syss_s *scr)
{
	if (!(cbc) || !(scr))
		return AILSA_NO_DATA;
	AILLIST *l = ailsa_db_data_list_init();
	AILLIST *res = ailsa_db_data_list_init();
	int retval;

	if ((retval = cmdb_add_string_to_list(scr->arg, l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add system script arg to list");
		goto cleanup;
	}
	if ((retval = cmdb_add_number_to_list(scr->no, l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add number to list");
		goto cleanup;
	}
	if ((retval = cmdb_add_system_script_id_to_list(scr->name, cbc, l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add system script id to list");
		goto cleanup;
	}
	if ((retval = cmdb_add_build_domain_id_to_list(scr->domain, cbc, l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add build domain id to list");
		goto cleanup;
	}
	if ((retval = cmdb_add_build_type_id_to_list(scr->type, cbc, l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add build type id to list");
		goto cleanup;
	}
	if ((retval = ailsa_argument_query(cbc, SYSTEM_SCRIPT_ARG_ID, l, res)) != 0) {
		ailsa_syslog(LOG_ERR, "SYSTEM_SCRIPT_ARG_ID query failed");
		goto cleanup;
	}
	if (res->total == 0) {
		ailsa_syslog(LOG_INFO, "System script arg %s for script %s not in database", scr->arg, scr->name);
		goto cleanup;
	}
	if ((retval = ailsa_delete_query(cbc, delete_queries[DELETE_SYSTEM_SCRIPT_ARG], res)) != 0)
		ailsa_syslog(LOG_INFO, "DELETE_SYSTEM_SCRIPT_ARG query failed");
	cleanup:
		ailsa_list_full_clean(l);
		ailsa_list_full_clean(res);
		return retval;
}

static int
cbc_script_list_script(ailsa_cmdb_s *cbc)
{
	if (!(cbc))
		return AILSA_NO_DATA;
	int retval;
	AILLIST *l = ailsa_db_data_list_init();
	AILELEM *e;

	if ((retval = ailsa_basic_query(cbc, SYSTEM_SCRIPT_NAMES, l)) != 0) {
		ailsa_syslog(LOG_ERR, "SYSTEM_SCRIPT_NAMES query failed");
		goto cleanup;
	}
	if (l->total == 0) {
		ailsa_syslog(LOG_INFO, "No system scripts to display");
		goto cleanup;
	}
	e = l->head;
	while (e) {
		printf("%s\n", ((ailsa_data_s *)e->data)->data->text);
		e = e->next;
	}
	cleanup:
		ailsa_list_full_clean(l);
		return retval;
}

static int
cbc_script_display_script(ailsa_cmdb_s *cbc, cbc_syss_s *scr)
{
	int retval = 0;

	if (!(scr->domain) && (scr->name))
		retval = cbc_script_args_display_all_domain(cbc, scr);
	else if ((scr->domain) && !(scr->name))
		retval = cbc_script_args_display_one_domain(cbc, scr);
	else
		retval = cbc_script_args_display_one_script(cbc, scr);
	return retval;
}

static int
cbc_script_args_display_all_domain(ailsa_cmdb_s *cbc, cbc_syss_s *scr)
{
	if (!(cbc) || !(scr))
		return AILSA_NO_DATA;
	AILLIST *l = ailsa_db_data_list_init();
	AILLIST *s = ailsa_db_data_list_init();
	AILELEM *e;
	size_t total = 4;
	char *domain = NULL, *alias = NULL;
	int retval, flag = 0;

	if ((retval = cmdb_add_string_to_list(scr->name, l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add script name to list");
		goto cleanup;
	}
	if ((retval = ailsa_argument_query(cbc, SYSTEM_SCRIPTS_ON_NAME, l, s)) != 0) {
		ailsa_syslog(LOG_ERR, "SYSTEM_SCRIPTS_ON_NAME query failed");
		goto cleanup;
	}
	if (s->total == 0) {
		ailsa_syslog(LOG_INFO, "No system script in database with name %s", scr->name);
		goto cleanup;
	}
	if ((s->total % total) != 0) {
		ailsa_syslog(LOG_INFO, "Got wrong number of results. Got %zu, wanted multiple of %zu", s->total, total);
		goto cleanup;
	}
	e = s->head;
	while (e) {
		flag = 0;
		if (!(domain)) 
			printf("Domain: %s", ((ailsa_data_s *)e->data)->data->text);
		else if (strcmp(domain, ((ailsa_data_s *)e->data)->data->text) != 0)
			printf("\n\nDomain: %s", ((ailsa_data_s *)e->data)->data->text);
		domain = ((ailsa_data_s *)e->data)->data->text;
		if (!(alias)) {
			printf("\n %s build\n", ((ailsa_data_s *)e->next->data)->data->text);
			flag = 1;
		} else if (strcmp(alias, ((ailsa_data_s *)e->next->data)->data->text) != 0) {
			printf("\n %s build\n", ((ailsa_data_s *)e->next->data)->data->text);
			flag = 1;
		}
		alias = ((ailsa_data_s *)e->next->data)->data->text;
		if (flag == 1)
			printf("\t%s", scr->name);
		printf(" %s", ((ailsa_data_s *)e->next->next->next->data)->data->text);
		e = ailsa_move_down_list(e, total);
	}
	printf("\n");
	cleanup:
		ailsa_list_full_clean(l);
		ailsa_list_full_clean(s);
		return retval;
}

static int
cbc_script_args_display_one_domain(ailsa_cmdb_s *cbc, cbc_syss_s *scr)
{
	if (!(cbc) || !(scr))
		return AILSA_NO_DATA;
	int retval, flag;
	size_t total = 4;
	char *script = NULL, *build = NULL;
	AILLIST *l = ailsa_db_data_list_init();
	AILLIST *s = ailsa_db_data_list_init();
	AILELEM *e;

	if ((retval = cmdb_add_string_to_list(scr->domain, l)) != 0) {
		ailsa_syslog(LOG_ERR, "SYSTEM_SCRIPTS_ON_DOMAIN query failed");
		goto cleanup;
	}
	if ((retval = ailsa_argument_query(cbc, SYSTEM_SCRIPTS_ON_DOMAIN, l, s)) != 0) {
		ailsa_syslog(LOG_ERR, "SYSTEM_SCRIPTS_ON_DOMAIN query failed");
		goto cleanup;
	}
	if (s->total == 0) {
		ailsa_syslog(LOG_INFO, "No system scripts configured for domain %s\n", scr->domain);
		goto cleanup;
	}
	if ((s->total % total) != 0) {
		ailsa_syslog(LOG_INFO, "Got wrong number of results. Got %zu, wanted mutiple of %zu", s->total, total);
		goto cleanup;
	}
	e = s->head;
	printf("Domain: %s\n", scr->domain);
	while (e) {
		flag = 0;
		if (!(build)) {
			printf(" %s build\n", (((ailsa_data_s *)e->next->data)->data->text));
			flag = 1;
		} else if (strcmp(build, ((ailsa_data_s *)e->next->data)->data->text) != 0) {
			printf("\n %s build\n", (((ailsa_data_s *)e->next->data)->data->text));
			flag = 1;
		}
		build = ((ailsa_data_s *)e->next->data)->data->text;
		if (flag == 1)
			printf("\t%s", ((ailsa_data_s *)e->data)->data->text);
		else if (!(script))
			printf("\t%s", ((ailsa_data_s *)e->data)->data->text);
		else if (strcmp(script, ((ailsa_data_s *)e->data)->data->text) != 0)
			printf("\n\t%s", ((ailsa_data_s *)e->data)->data->text);
		script = ((ailsa_data_s *)e->data)->data->text;
		printf(" %s", ((ailsa_data_s *)e->next->next->next->data)->data->text);
		e = ailsa_move_down_list(e, total);
	}
	printf("\n");
	cleanup:
		ailsa_list_full_clean(l);
		ailsa_list_full_clean(s);
		return retval;
}

static int
cbc_script_args_display_one_script(ailsa_cmdb_s *cbc, cbc_syss_s *scr)
{
	if (!(cbc) || !(scr))
		return AILSA_NO_DATA;
	AILLIST *l = ailsa_db_data_list_init();
	AILLIST *s = ailsa_db_data_list_init();
	AILELEM *e;
	char *build = NULL;
	int retval, flag;
	size_t total = 3;

	if ((retval = cmdb_add_string_to_list(scr->name, l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add script name to list");
		goto cleanup;
	}
	if ((retval = cmdb_add_string_to_list(scr->domain, l)) != 0) {
		ailsa_syslog(LOG_ERR, "Cannot add build domain to list");
		goto cleanup;
	}
	if ((retval = ailsa_argument_query(cbc, SYSTEM_SCRIPTS_ON_NAME_DOMAIN, l, s)) != 0) {
		ailsa_syslog(LOG_ERR, "SYSTEM_SCRIPTS_ON_NAME_DOMAIN query failed");
		goto cleanup;
	}
	if (s->total == 0) {
		ailsa_syslog(LOG_INFO, "No script configuration found for %s in build domain %s\n", scr->name, scr->domain);
		goto cleanup;
	}
	if ((s->total % total) != 0) {
		ailsa_syslog(LOG_INFO, "Got wrong number of results. Got %zu, wanted mutiple of %zu", s->total, total);
		goto cleanup;
	}
	e = s->head;
	printf("Domain: %s\n", scr->domain);
	while (e) {
		flag = 0;
		if (!(build)) {
			flag = 1;
			printf(" %s build\n", (((ailsa_data_s *)e->data)->data->text));
		} else if (strcmp(build, (((ailsa_data_s *)e->data)->data->text)) != 0) {
			flag = 1;
			printf("\n %s build\n", (((ailsa_data_s *)e->data)->data->text));
		}
		build = ((ailsa_data_s *)e->data)->data->text;
		if (flag == 1)
			printf("\t%s", scr->name);
		printf(" %s", (((ailsa_data_s *)e->next->next->data)->data->text));
		e = ailsa_move_down_list(e, total);
	}
	printf("\n");
	cleanup:
		ailsa_list_full_clean(l);
		ailsa_list_full_clean(s);
		return retval;
}
